lib/galaxy/workflow/modules.py

```diff
@@ -481,9 +481,8 @@ def get_all_outputs(self, data_only=False):
                 step = subworkflow_dict['steps'][order_index]
                 data_outputs = step['outputs']
                 for workflow_output in step['workflow_outputs']:
-                    label = workflow_output['label']
-                    if not label:
-                        label = "%s:%s" % (order_index, workflow_output['output_name'])
+                    label = workflow_output['label'] or workflow_output['output_name']
+                    label = "%s:%s" % (order_index, label)
```

Comment:
I think this is intentional, the order_index is not necessary stable, and the label is important to identify connections.
As is this would probably break existing workflows. What is the usecase you're trying to enable ?

Reply:
I should add that added `label = "%s:%s" % (order_index, workflow_output['output_name'])` as a fallback to make it more obvious how to use outputs from a workflow, since the requirement to label them wasn't (and still isn't) exposed in the UI. I am still not 100% correct that this was the right choice.

Reply:
Thanks for reviewing this. The bug I observe is reproducible when a 3-fold nested subworkflow is created. I created a 1st workflow containing a simple tool like `Add Column`, then imported it as a subworkflow into a 2nd workflow and finally imported the 2nd workflow as subworkflow into a 3rd workflow. When importing the 1st into the 2nd workflow labels are added to the workflow outputs and when the 2nd workflow is imported into the 3rd workflow those labels remain unchanged i.e. additional indices are not added.

The fixed labels for subworkflows are added at: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/workflow/modules.py#L491.
I am working on a fix. I think the above condition might be the issue. Possible related to the uuid handling for workflow outputs.

Reply:
I dug a little bit deeper and noticed that the underlying reason for this issue is that the new client accepts and preserves subworkflow output labels if they are specified in the backend (as it should). As a consequence, the workflow output label is not further augmented by adding the step number if the workflow is nested again. In contrast, the previous workflow displayed the subworkflow output label but did not parse it back when the workflow was saved. I adjusted this PR to fix this issue and restore the equivalent behavior by not parsing the label to the client. The client will fallback on the name which contains the same value without having to explicitly store a label.